examples/calibration/plot_compare_calibration.py

```diff
@@ -150,61 +158,114 @@ def predict_proba(self, X):
 plt.show()
 
 # %%
-# :class:`~sklearn.linear_model.LogisticRegression` returns well calibrated
-# predictions as it directly optimizes log-loss. In contrast, the other methods
-# return biased probabilities, with different biases for each method:
 #
-# * :class:`~sklearn.naive_bayes.GaussianNB` tends to push
-#   probabilities to 0 or 1 (see histogram). This is mainly
-#   because the naive Bayes equation only provides correct estimate of
+# Analysis of the results
+# -----------------------
+#
+# :class:`~sklearn.linear_model.LogisticRegressionCV` returns reasonably well
+# calibrated predictions despite the small training set size: its reliability
+# curve is the closest to the diagonal among the four models.
+#
+# Logistic regression is trained by minimizing the log-loss which is a strictly
+# proper scoring rule: in the limit of infinite training data, strictly proper
```

Comment:
We are getting closer;-)

I'm a bit undecided whether or not to mention asymptomatic arguments. We don’t need to, the expectation of proper scoring rules is minimized by the true probabilities. The asymptotic argument comes in only when estimating this expectation.

Note that also the RF uses proper scoring rules, e.g. Gini criterion is the squared error evaluated with the mean of that node as prediction.
I guess that it is more the overfitting difference that we see.

Arrrg, difficult to keep it simple.

Reply:
> I'm a bit undecided whether or not to mention asymptomatic arguments. We don’t need to, the expectation of proper scoring rules is minimized by the true probabilities. The asymptotic argument comes in only when estimating this expectation.

I think it's important because most of the time machine learning practitioners are interested in out-of-sample predictions.

> Note that also the RF uses proper scoring rules, e.g. Gini criterion is the squared error evaluated with the mean of that node as prediction.

Yes, I was thinking about updating the item about the RF but was worried to make it too complex.

> The fact that LR uses a canonical loss-link combination is also very important for the calibration results.

I can try to fit this somewhere, but since do not really expose custom link functions for classifiers, I am not sure if this will make the message too complex without justification in terms of choice of hyperparameters.

Reply:
I expanded the sections on RFs. I also conducted experiments on synthetic data. I confirm I can reproduce the effect reported by Caruana et al. (by playing with `max_features` and a dataset with a large number of uninformative features) so it's not just a theoretical artifact.

However, RF can also be strongly over-confident if the number of trees is not large enough to counteract the natural overfitting/over-confident behavior of a small number of unbounded trees. So the calibration picture can be really complex for RFs.